apps/sms/test/marionette/mocks/mock_navigator_moz_mobile_message.js

```diff
@@ -27,215 +24,210 @@ Services.obs.addObserver(function(document) {
-          cursor.onsuccess.call(cursor);
+        if (!iteratorResult.done) {
+          Services.DOMRequest.fireSuccessAsync(cursor, iteratorResult.value);
+        } else {
+          Services.DOMRequest.fireDone(cursor);
         }
-      }, delayMs);
+      };
+
+      var cursor = Services.DOMRequest.createCursor(window, handleCursor);
+
+      getStorage().then(function(storage) {
+        threads = storage.threads.values();
+
+        handleCursor();
+      });
 
       return cursor;
     },
 
     getMessages: function(filter) {
-      var cursor = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null,
-        result: null,
-        continue: null,
-        done: false
-      }, window));
+      var messages = null;
 
-      window.setTimeout(function() {
-        if (typeof cursor.onsuccess === 'function') {
-          var thread = threads.get(filter.threadId),
-              currentIndex = -1;
-
-          cursor.continue = function() {
-            if (!thread || ++currentIndex >= thread.messages.length) {
-              cursor.done = true;
-              cursor.continue = null;
-              cursor.result = null;
-            } else {
-              cursor.result = thread.messages[currentIndex];
-            }
-            cursor.onsuccess.call(cursor);
-          };
+      var handleCursor = function() {
+        var iteratorResult = messages.next();
 
-          cursor.continue();
+        if (!iteratorResult.done) {
+          Services.DOMRequest.fireSuccessAsync(cursor, iteratorResult.value);
+        } else {
+          Services.DOMRequest.fireDone(cursor);
         }
-      }, delayMs);
+      };
+
+      var cursor = Services.DOMRequest.createCursor(window, handleCursor);
+
+      getStorage().then(function(storage) {
+        var thread = storage.threads.get(filter.threadId);
+
+        // Remove this once Array.prototype.values() is landed (see bug 875433).
+        messages = new Set(thread && thread.messages || []).values();
+
+        handleCursor();
+      });
 
       return cursor;
     },
 
     getMessage: function(id) {
-      var request = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null,
-        result: null
-      }, window));
-
-      window.setTimeout(function() {
-        if (typeof request.onsuccess === 'function') {
-          threads.forEach(function(thread) {
-            thread.messages.forEach(function(message) {
-              if (message.id === id) {
-                request.result = message;
-              }
-            });
+      var request = Services.DOMRequest.createRequest(window);
+
+      getStorage().then(function(storage) {
+        var requestedMessage = null;
+        storage.threads.forEach(function(thread) {
+          thread.messages.forEach(function(message) {
+            if (message.id === id) {
+              requestedMessage = message;
+            }
           });
-          request.onsuccess.call(request);
+        });
```

Comment:
maybe the opposite makes more sense: the test would send a "messages" array and setStorage would organize this into threads. This would make a simpler test. But maybe it's more readable to concretely show the thread in the test. Tell me what you think :)